Add documentation file for TS/TSX (#226)
MaibornWolff/metric-gardener: TS_TSX.md

# File Metrics for TypeScript (TS) and TypeScript JSX (TSX)

This document outlines the definition of file metrics for TypeScript (TS) and TypeScript JSX (TSX).

> ### 1. classes

Reviewer: We do not have numbers for them in the main Readme. So I think, we should not do this here, either. Also, the order of the metrics here does not match that of the readme. We should probably order them identical to the readme here.

The "classes" metric counts:

> -   keywords like: `class`, `abstract class`, `interface`, `enum`

Reviewer: Is it correct to say that we really count the keyword? Technically, we do not, as we count the nodes `class_declaration` and not the (unnamed) node that contains the keyword itself. 
Practically, I don't know if the keywords `class`, etc. can be only used to define a class in TS or also on other places, so that that your documentation is right anyway or not. For C and C++, this would be clearly wrong, as we can do things like 
````C++
#include <iostream>

class class_name;

class class_name {
    public:
        void print_hello() {
            std::cout << "Hello!" << std::endl;
        }
};

int main() {
    class class_name* object = new class_name();
    class class_name object2();
    class class_name object3();

    object->print_hello();
    object2.print_hello();
    object3.print_hello();
}
````

Here, we actually do only count the thing with the second class keyword.

Reviewer: Also, what does "keywords like" mean? Do we also count other keywords that are similar? Then we should mention them here.

I would not write here that we count keywords at all, however. We should write that we count class declarations, abstract class declarations, interface declarations and enum declarations.

> -   type aliases with object type literals e.g. `type ObjectType = { message: string};`

Reviewer: As we currently do not count them when there is more than one object type literal:
```suggestion
-   type aliases with a single object type literal, e.g. `type ObjectType = { message: string};`
```

It does **not** count:

-   other type aliases like
    -   `type Union = {a: 1} & {b: 2};`
    -   `type Intersect = {a: 1} | {b: 2};`
    -   `type Tuple = [name: string, age: number];`
    -   `type AliasType = number;`

### 2. functions

The "functions" metric counts:

> -   methods declarations in classes, abstract classes and interfaces

Reviewer: ```suggestion
-   method declarations in classes, abstract classes and interfaces
```

> -   static methods, constructors, getters and setters

Reviewer: Whats with the static initialization block you mentioned recently?

-   functions declared with the keyword `function`
-   arrow functions

### 3. complexity

The "complexity" metric counts:

> -   `if`, `if else` statements, but **no** `else` statements

Reviewer: I don't know if we should explicitely mention that we count no `else` here, as that is already clear when reading the metric definition in the readme and when mentioning that we count for `if` and `if else`.

Reviewer: The same holds for `default` and `finally` below

-   ternary expressions

> -   for, while, do-while loops

Reviewer: We should write `do..while` just like they do here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/do...while?retiredLocale=de

Also a nit:  using "and" instead of a comma before the last element of the list is a better writing style.
```suggestion
-   `for`, `while` and `do...while` loops
```

Reviewer: We should probably also mention `for...of`, `for await...of` and `for...in` loops here.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in

-   `case` labels in switch-statements, but **no** `default` labels
-   `catch` labels in try-catch blocks, but **no** `finally` labels

> -   binary operations `&&` and `||`

Reviewer: Like in the readme, we should make clear that we only count **logical** binary operations, not things like `+`, `-`, `*`, `/`, etc.
```suggestion
-   logical binary operations `&&` and `||`
```

> -   everything counted for "functions" metric

Reviewer: ```suggestion
-   everything counted for the "functions" metric
```

It does **not** count:

-   `.catch()` method call for Promises

> -   `.forEach()` method call for iterables

Reviewer: There are also similar functions like `map`, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map. We should make clear that do not count them as well:
```suggestion
-   `.forEach()` method call for iterables and similar build-in functions
```

This document outlines the definition of file metrics for TypeScript (TS) and TypeScript JSX (TSX).

### complexity

The "complexity" metric counts:

-   `if` and `if else` statements
-   ternary expressions
-   `for...of`, `for await...of`, `for...in`,`while` and `do...while` loops
-   `case` labels in switch-statements

> -   `catch` labels in try-catch blocks

Reviewer: They are no label inside a try, but a block themself.
```suggestion
-   `catch` blocks
```

-   logical binary operations `&&` and `||`
-   everything counted for the "functions" metric

It does **not** count:

-   `.catch()` method call for Promises
-   `.forEach()` method call for iterables and other built-in functions

### functions

The "functions" metric counts:

-   method declarations in classes, abstract classes and interfaces
-   static methods, constructors, getters and setters
-   functions declared with the keyword `function`
-   arrow functions

### classes

The "classes" metric counts:

> -   declarations of class, abstract class, interface and enum

Reviewer: Should be in plural
```suggestion
-   declarations of classes, abstract classes, interfaces and enums
```

-   type aliases declaration with one single object type literal e.g. `type ObjectType = { message: string};`

It does **not** count:

-   other type aliases like
    -   `type Union = {a: 1} & {b: 2};`
    -   `type Intersect = {a: 1} | {b: 2};`
    -   `type Tuple = [name: string, age: number];`
    -   `type AliasType = number;`

### lines_of_code, comment_lines and real_lines_of_code

> see README.md

Reviewer: That would be more cool
